packages/ariakit/src/popover/popover-state.ts

```diff
@@ -128,105 +125,90 @@ export function usePopoverState({
-              // move it up when the placement is "right-start". A good example
-              // is a sub menu that must have a small negative shift so the
-              // first menu item is aligned with its menu button.
-              offset: ({ placement }: { placement: Placement }) => {
-                const start = placement.split("-")[1] === "start";
-                return [start ? shift : -shift, finalGutter];
-              },
-            },
-          },
-          {
-            // https://popper.js.org/docs/v2/modifiers/prevent-overflow/
-            name: "preventOverflow",
-            enabled: preventOverflow && dialog.mounted,
-            options: {
-              padding,
-              tetherOffset: finalGutter,
-            },
-          },
-          {
-            // https://popper.js.org/docs/v2/modifiers/arrow/
-            name: "arrow",
-            enabled: !!arrow,
-            options: { element: arrow, padding: arrowPadding },
-          },
-          {
-            // https://codesandbox.io/s/bitter-sky-pe3z9?file=/src/index.js
-            name: "sameWidth",
-            enabled: sameWidth,
-            phase: "beforeWrite",
-            requires: ["computeStyles"],
-            fn: ({ state }) => {
-              if (state.styles.popper) {
-                state.styles.popper.width = `${state.rects.reference.width}px`;
-              }
-            },
-            effect: ({ state }) => {
-              const { reference } = state.elements;
-              const referenceElement = (
-                "contextElement" in reference
-                  ? reference.contextElement
-                  : reference
-              ) as HTMLElement | null;
-              if (referenceElement && "offsetWidth" in referenceElement) {
-                const referenceWidth = referenceElement.offsetWidth + "px";
-                state.elements.popper.style.width = referenceWidth;
+      const update = async () => {
+        if (!dialog.mounted) return;
+
+        const middleware: Middleware[] = [
+          middlewares.offset(({ placement }) => {
+            // Makes sure the shift value is applied to the popover element
+            // consistently no matter the placement. That is, a negative shift
+            // should move down a popover with a "right-end" placement, but move
+            // it up when the placement is "right-start". A good example is a
+            // sub menu that must have a small negative shift so the first menu
+            // item is aligned with its menu button.
+            const start = placement.split("-")[1] === "start";
+            return {
+              crossAxis: start ? shift : -shift,
+              mainAxis: finalGutter,
+            };
+          }),
+        ];
+
+        if (preventOverflow) {
+          middleware.push(middlewares.shift({ padding }));
+        }
+
+        if (flip) {
+          middleware.push(middlewares.flip({ padding }));
+        }
```

Comment:
Usually `flip` should go before `shift` (although, there may be exceptions sometimes)

https://floating-ui.com/docs/flip#flipalignment